Initial work on code cleanup

sebastiano-barrera opened the pull request:
Here's a (hopefully complete) list of the changes in this commit:
- The `HeightmapRequest` message has been removed: it was a placeholder
  and it's been replaced by `heightmap_node`'s `Query` service.
  Also, `heightmap` has been added, as a dependency, in
  `CMakeLists.txt`.
- The node is now compatible with `heightmap`.
- The `constants` module has been simplified.
- The `debug_draw` has been moved to a dedicated module, which is
  currently unused.
- The default Python version has been changed, sadly, from 3 to 2: there
  are some headaches to solve before being able to use Python 3 properly
  with ROS (it will be done eventually, though)
- Removed a bunch of superfluous `import` statements.
- The `ToFSuspensionController` class has been heavily simplified.
- Managing errors by checking return values is un-pythonic, and really
  hurts readability of the code. The Python way is using exceptions, and
  managing exceptions only in those places where it makes sense to do
  so (which ROS already does by default, so we can just remove most error
  management code and be done with it).
- Some more Python idioms were used to make some pieces of code more
  readable, less repetitive.

sebastiano-barrera committed:
Initial work on code cleanup

Here's a (hopefully complete) list of the changes in this commit:

* The `HeightmapRequest` message has been removed: it was a placeholder
  and it's been replaced by ``heightmap_node``'s ``Query`` service.
  Also, ``heightmap`` has been added, as a dependency, in
  ``CMakeLists.txt``.

* The node is now compatible with ``heightmap``.

* The ``constants`` module has been simplified.

* The ``debug_draw`` has been moved to a dedicated module, which is
  currently unused.

* The default Python version has been changed, sadly, from 3 to 2: there
  are some headaches to solve before being able to use Python 3 properly
  with ROS (it will be done eventually, though)

* Removed a bunch of superfluous ``import`` statements.

* The ``ToFSuspensionController`` class has been heavily simplified.

* Managing errors by checking return values is un-pythonic, and really
  hurts readability of the code. The Python way is using exceptions, and
  managing exceptions only in those places where it makes sense to do
  so (which ROS already does by default, so we can just remove most error
  management code and be done with it).

* Some more Python idioms were used to make some pieces of code more
  readable, less repetitive.

tamersaadeh commented on src/suspension_controller/tof_suspension.py:
```diff
@@ -9,104 +7,70 @@
+        rospy.loginfo("Started")
 
-    def __unpublish(self):
+    def unregister(self):
```
This was done on purpose as it should not be visible outside this class (same applied to the `__publish()` function)

sebastiano-barrera replied:
Agreed (this method is never called, so I didn't know what to think of it when I first read the code). But the Python convention is single underscore, because double underscore has a special meaning for the interpreter.

tamersaadeh replied:
isn't the special meaning that it is private?

sebastiano-barrera replied:
Ehh... not really. Python rejects member visibility on purpose (a principle the Python community calls "Python is a language for consenting adults"), the single underscore is just a convention. Names starting with double-underscore have special rules. See here for a nice summary: http://stackoverflow.com/questions/1301346/the-meaning-of-a-single-and-a-double-underscore-before-an-object-name-in-python

tamersaadeh replied:
Exactly the meaning I was looking for: "class private" function!!!! Because the idea is that I wanted to not have the function called externally but not write all that code in the constructor. I did that because it is easier later when I can merge the range finder stuff as the publish/unpublish pair would be much more complex than the one I'm currently using. 

> On 24 Nov 2015, at 16:46, sebastiano-barrera notifications@github.com wrote:
> 
> In src/suspension_controller/tof_suspension.py:
> 
> > -    def __unpublish(self):
> > -    def unregister(self):
> >   Ehh... not really. Python rejects member visibility on purpose (a principle the Python community calls "Python is a language for consenting adults"), the single underscore is just a convention. Names starting with double-underscore have special rules. See here for a nice summary: http://stackoverflow.com/questions/1301346/the-meaning-of-a-single-and-a-double-underscore-before-an-object-name-in-python
> 
> —
> Reply to this email directly or view it on GitHub.

tamersaadeh replied:
I just realised that I want it to be private but I still pass it to ros, so it has to be public!

tamersaadeh commented on src/suspension_controller/tof_suspension.py:
```diff
@@ -9,104 +7,70 @@
 
+    ## XXX: check these numbers with the rover
+    WHEEL_NAMES = ['f_r', 'f_l', 'b_r', 'b_l']
+
```
NO this is WIP when I add vincenzo's node (this is the exact code I was talking about)

tamersaadeh replied:
fixed

tamersaadeh commented on src/suspension_controller/tof_suspension.py:
```diff
@@ -9,104 +7,70 @@
 
     def get_heightmaps(self):
-        heightmaps = []
+        return [self.get_heightmap(i) for i in (1, 2, 3, 4)]
```
but it should check if it has an error, that's why I didn't use the simplified version! So unless you can do that in a one-liner, I think we should stick to the previous version

tamersaadeh commented on src/suspension_controller/tof_suspension.py:
```diff
@@ -9,104 +7,70 @@
+                                       y_size=constants.ARM_LENGTH,
+                                       x_samples=1,
+                                       y_samples=1000)
+        return np.array(res.map).reshape(res.y_samples, res.x_samples)
```
you can do that directly?! In any case that's better

tamersaadeh commented on src/suspension_controller/tof_suspension.py:
```diff
@@ -9,104 +7,70 @@
-        return heightmaps
-
-    def __find_theta(self, height, hmap, eps=0.05):
+    def _find_theta(self, height, hmap, eps=0.05):
```
see comment on `__unpublish()` and `__publish()`

tamersaadeh commented on src/suspension_controller/tof_suspension.py:
```diff
@@ -155,85 +119,37 @@ def find_legs_solution(self, heightmaps, eps=0.05):
+                  .format(task_description, elapsed_time.secs))
+
 
+## TODO add solution for 4 arms
```
What do you mean?! It's already in the code! See `get_heightmaps()`

sebastiano-barrera replied:
This is not my comment, I just transcribed it

tamersaadeh replied:
ok, I thought I dropped it

tamersaadeh commented on src/suspension_controller/constants.py:
```diff
@@ -6,20 +6,9 @@
 # Copyright (c) 2014-2015, Tamer Saadeh <tamer@tamersaadeh.com>
 # All rights reserved.
 
-class _const:
```
The main reason I did this was so that the constants don't leak into the global namespace of variables/functions in python, I think it is better if it can be simplified without pollution the namespace (eg yaml), but there's not much time to do so.

sebastiano-barrera replied:
That's the whole point of modules, though... Just don't `from constants import *`

tamersaadeh replied:
ok

sebastiano-barrera committed:
continuing cleanup: small fixes

This commit contains the following bug fixes:

* `measure_time` looked for some variables by calling them with the wrong
  name (`later` instead of `end_time` and `now` instead of `init_time`)

* In `main`: `rate` was assigned the wrong value (an extra call to
  `sleep` that shouldn't have been there)

* In `ToFSuspensionController._find_theta`: references to
  `self.req_height` should have been to the `height` argument of the
  method.

* Wheels are no longer indicated by their indices, but by the name of
  their TF transform. This is still very much a temporary solution, and
  not very different from the previous, but it's slightly cleaner.

tamersaadeh commented:
Ok this looks better, could you split it into multiple smaller diffs as I can cherry-pick some instead of all of them?

sebastiano-barrera commented:
I can't really do much from my side. I think you have to do it yourself (`git cherry-pick` might just do the trick)

tamersaadeh commented:
Well it's all in a single commit,  but I can manually choose them if that's ok with you. But we lose authorship I could reference the pull though... 

> On 24 Nov 2015, at 16:49, sebastiano-barrera notifications@github.com wrote:
> 
> I can't really do much from my side. I think you have to do it yourself (git cherry-pick might just do the trick)
> 
> —
> Reply to this email directly or view it on GitHub.
